### PR TITLE
Add methods for keychains backup and restore

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Now it will be possible to backup and selectively restore keychains during full reset

### CODE REVIEW

File: lib/simulator-xcode-6.js [hunk 1]
```diff
@@ -951,6 +951,68 @@ class SimulatorXcode6 extends EventEmitter {
     `);
   }
 
+  //region Keychains Interaction
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: is this a standard thing or is it for a particular IDE?

File: lib/simulator-xcode-6.js [hunk 2]
```diff
@@ -951,6 +951,68 @@ class SimulatorXcode6 extends EventEmitter {
...
+  async backupKeychains () {
+    if (!await fs.exists(this.keychainPath)) {
+      return false;
+    }
+
+    const backupPath = await tempDir.path({
+      prefix: `keychains_backup_${Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1)}`,
+      suffix: '.zip',
+    });
+    const cmdLine = ['-r', backupPath, `${this.keychainPath}${path.sep}`];
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: `const zipArgs`

File: lib/simulator-xcode-6.js [hunk 3]
```diff
@@ -951,6 +951,68 @@ class SimulatorXcode6 extends EventEmitter {
...
+    log.debug(`Creating keychains backup with 'zip ${cmdLine.join(' ')}' command`);
+    await exec('zip', cmdLine);
+    if (_.isString(this._keychainsBackupPath) && await fs.exists(this._keychainsBackupPath)) {
+      await fs.unlink(this._keychainsBackupPath);
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: why use `unlink` here and `rimraf` below? is there a perf difference?

File: lib/simulator-xcode-6.js [hunk 4]
```diff
@@ -951,6 +951,68 @@ class SimulatorXcode6 extends EventEmitter {
...
+      throw new Error(`The keychains backup archive does not exist. ` +
+                      `Are you sure it was created before?`);
+    }
+
+    if (_.isString(excludePatterns)) {
+      excludePatterns = excludePatterns.split(',').map((x) => x.trim());
+    }
+    await fs.rimraf(this.keychainPath);
+    await mkdirp(this.keychainPath);
+    const cmdLine = ['-o', this._keychainsBackupPath,
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: `const unzipArgs`

File: lib/simulator-xcode-6.js [hunk 5]
```diff
@@ -951,6 +951,68 @@ class SimulatorXcode6 extends EventEmitter {
...
+                     ...(_.flatMap(excludePatterns.map((x) => ['-x', x]))),
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: ooooo spread and flatmap 😍

Comment (mykola-mokhnach) [reply lib/simulator-xcode-6.js]: yeah, functional programming rocks

File: lib/simulator-xcode-6.js (hunk 3 above)
Comment (mykola-mokhnach) [reply lib/simulator-xcode-6.js]: will change to unlink, since it is anyway a single file

File: lib/simulator-xcode-6.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/simulator-xcode-6.js]: I assume nowadays it is supported by most of IDEs

Comment (mykola-mokhnach) [reply lib/simulator-xcode-6.js]: at least Eclipse, IDEA and VSCode

**Pull request merged**